[Merged by Bors] - feat(RingTheory/AdicCompletion): refactor and expand API

chrisflav opened the pull request:
Refactors and expands the API for adic completions. More precisely:

- Renames `adicCompletion` to `AdicCompletion`.
- Redefines `adicCompletion` in terms of a `transitionMap`.
- Introduces `AdicCauchySequence` as the module of `I`-adic Cauchy sequences.
- Provides a surjective linear map `AdicCauchySequence I M →ₗ[R] AdicCompletion I M` and an induction principle for `AdicCompletion I M` for that.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->
- [x] depends on: #12511 

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -83,18 +83,27 @@ def Hausdorffification : Type _ :=
+/-- The completion of a module with respect to an ideal. This is not necessarily Hausdorff.
+In fact, this is only complete if the ideal is finitely generated. -/
+def AdicCompletion : Type _ :=
+  adicCompletion I M
```
Maybe we can define this to be `{ f // ∀ {m n} (hmn : m ≤ n), AdicCompletion.transitionMap I M hmn (f n) = f m }` and ditch `adicCompletion` entirely? Similarly for `adicCauchySequence` (maybe replace it with a predicate `IsAdicCauchy`)

chrisflav replied:
How do I then get the `AddCommGroup` and `Module` instances? Do I have to write them by hand? Currently they are inherited by `adicCompletion` being a `Submodule`.

chrisflav replied:
And I have realized that `adicCompletion` should not be private after all, since it comes in handy again for defining the algebra instance. Maybe rename `adicCompletion` in `submodule` and define `AdicCompletion` as you suggest. Then I think `inferInstanceAs <| AddCommGroup (submodule I M)` should still work.

erdOne replied:
I suppose the proof that pointwise add and smul works is exactly the same proof as showing `adicCompletion` is a submodule. But if you will need it elsewhere then I think your suggestion is the way to go.

chrisflav replied:
I think I changed it according to your suggestions.

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +210,204 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+
+@[simp]
+theorem transitionMap_comp {m n k : ℕ} (hmn : m ≤ n) (hnk : n ≤ k) :
+    transitionMap I M hmn ∘ₗ transitionMap I M hnk = transitionMap I M (hmn.trans hnk) := by
```
Can you add `(transitionMap I M hmn (transitionMap I M hnk x)) = transitionMap I M (hmn.trans hnk) x` as a simp lemma?

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +210,204 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+
+@[simp]
+theorem transitionMap_comp_eval {m n : ℕ} (hmn : m ≤ n) :
+    transitionMap I M hmn ∘ₗ eval I M n = eval I M m := by
```
Same here.

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +210,204 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+  toFun := fun x ↦ ⟨fun n ↦ f n x, by
+      intro k l hkl
+      change ((transitionMap I N hkl) ∘ₗ (f l)) x = _
+      rw [h hkl]
+    ⟩
```
```suggestion
      rw [h hkl]⟩
```

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +203,228 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+theorem transitionMap_mk {m n : ℕ} (hmn : m ≤ n) (x : M) :
+    transitionMap I M hmn
+      (Submodule.Quotient.mk (p := (I ^ n • ⊤ : Submodule R M)) x) =
+      Submodule.Quotient.mk (p := (I ^ m • ⊤ : Submodule R M)) x := by
+  simp [transitionMap]
```
```suggestion
      Submodule.Quotient.mk (p := (I ^ m • ⊤ : Submodule R M)) x :=
   rfl
```

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +203,228 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+
+@[simp]
+theorem transitionMap_comp_eval_apply {m n : ℕ} (hmn : m ≤ n) (x : AdicCompletion I M) :
+    transitionMap I M hmn (x.val n) = x.val m :=
```
Which is the simp normal form? `eval I M n f` or `f.1 n`? You should choose one and design the simp lemmas around it. If it is the former, than this lemma should be stated with it. If it is the latter, `AdicCompletion.eval_apply` should be a simp lemma and `val_add`, `val_sub`, `val_smul`, `val_zero` simp lemmas should be added.

chrisflav replied:
The simp NF is `f.1 n` because of `AdicCompletion.coe_eval` being a simp lemma. That's why `AdicCompletion.eval_apply` can't be a simp lemma currently, because it is already proven by `simp`. Are you saying that `eval_apply` should be the simp lemma instead of `coe_eval`?

For the time being, I added the simp lemmas you suggested. Anyway, I am not entirely sure whether the simp NF should not be `eval I M n f` instead, because it unveils less of the internal implementation of `AdicCompletion`. It might also be easier to control in the `Algebra` case, since then there are two "canonical" projections. The one to `R ⧸ (I ^ n • ⊤)` (`eval I R n`) and the one to `R ⧸ I ^ n` (`evalₐ I n` which is defined in the Algebra PR).

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +203,228 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+        · refine SModEq.mono (smul_mono (Ideal.pow_le_pow_right hmn) (by rfl)) (h n)
+
+/-- Construct `I`-adic cauchy sequence from sequence satisfying the successive cauchy condition. -/
+def AdicCauchySequence.mk (f : ℕ → M)
```
```suggestion
@[simps] def AdicCauchySequence.mk (f : ℕ → M)
```
Maybe?

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +203,228 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+    intro r f hf m n hmn
+    exact SModEq.smul (hf hmn) r
+
+instance : AddCommGroup (AdicCauchySequence I M) :=
+  inferInstanceAs <| AddCommGroup (AdicCauchySequence.submodule I M)
+
+instance : Module R (AdicCauchySequence I M) :=
+  inferInstanceAs <| Module R (AdicCauchySequence.submodule I M)
```
Can you add simp lemmas such as `(f + g).1 x = f.1 x + f.2 x`?

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +203,228 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+  smul_mem' := by
+    intro r f hf m n hmn
+    exact SModEq.smul (hf hmn) r
+
```
Maybe `CoeFun (AdicCauchySequence I M) (ℕ → M)`?

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +203,228 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+/-- Every element in the adic completion is represented by a Cauchy sequence. -/
+theorem mk_surjective : Function.Surjective (mk I M) := by
+  intro x
+  have h (n : ℕ) :
+      ∃ (a : M), Submodule.Quotient.mk (p := (I ^ n • ⊤ : Submodule R M)) a = x.val n :=
+    Quotient.exists_rep (x.val n)
+  choose a ha using h
```
```suggestion
  choose a ha using fun n ↦ Submodule.Quotient.mk_surjective _ (x.val n)
```

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +203,228 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+  toFun f := ⟨fun n ↦ Submodule.mkQ (I ^ n • ⊤ : Submodule R M) (f.val n), by
+    intro m n hmn
+    simp only [mkQ_apply, transitionMap_mk]
+    exact (f.property hmn).symm
+    ⟩
```
```suggestion
    exact (f.property hmn).symm⟩
```

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +203,228 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+def lift (f : ∀ (n : ℕ), M →ₗ[R] N ⧸ (I ^ n • ⊤ : Submodule R N))
+    (h : ∀ {m n : ℕ} (hle : m ≤ n), transitionMap I N hle ∘ₗ f n = f m) :
+    M →ₗ[R] AdicCompletion I N where
+  toFun := fun x ↦ ⟨fun n ↦ f n x, by
+    intro k l hkl
+    change ((transitionMap I N hkl) ∘ₗ (f l)) x = _
+    rw [h hkl]⟩
```
```suggestion
  toFun := fun x ↦ ⟨fun n ↦ f n x, fun hkl ↦ LinearMap.congr_fun (h hkl) x⟩⟩
```

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +203,228 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+    (h : ∀ n, f n ≡ f (n + 1) [SMOD (I ^ n • ⊤ : Submodule R M)]) : AdicCauchySequence I M where
+  val := f
+  property := by
+    rw [isAdicCauchy_iff]
+    exact h
```
```suggestion
    rwa [isAdicCauchy_iff]
```

chrisflav commented:
Thanks a lot for all the suggestions @erdOne. I think I addressed them as you suggested.

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +203,259 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
-end adicCompletion
+@[simp]
+theorem transitionMap_mk {m n : ℕ} (hmn : m ≤ n) (x : M) :
+    transitionMap I M hmn
+      (Submodule.Quotient.mk (p := (I ^ n • ⊤ : Submodule R M)) x) =
+      Submodule.Quotient.mk (p := (I ^ m • ⊤ : Submodule R M)) x := by
```
```suggestion
    transitionMap I M hmn (Submodule.Quotient.mk x) =
      Submodule.Quotient.mk x := by
```
I think this works and looks cleaner, but the current statement is also fine if you prefer explicitness.

chrisflav replied:
Removing the explicit annotations makes this significantly slower (measured by eye). IMHO it might be a good idea to make `p` an explicit argument of `Submodule.Quotient.mk`, as the unification (at least that's what I think it is) seems to be quite slow there. Note that the ideal argument of `Ideal.Quotient.mk` is also explicit.

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +203,259 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+@[simp]
+theorem val_zero (n : ℕ) : (0 : AdicCauchySequence I M) n = 0 :=
+  rfl
+
```
Maybe a simp lemma `f.1 x = f x` if the latter is simp NF.

chrisflav replied:
This is proven by `simp`, should I add it anyway?

erdOne replied:
Oh maybe coefun is reducible or something similar.
I think the lemma might still be useful though?

chrisflav replied:
CI says "LHS equals RHS syntactically", so I removed it again.

erdOne replied:
Ah coefun is smarter than I think. Sorry for the confusion.

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +203,259 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+    rfl
+
+@[simp]
+lemma lift_eval (f : ∀ (n : ℕ), M →ₗ[R] N ⧸ (I ^ n • ⊤ : Submodule R N))
```
```suggestion
lemma eval_lift (f : ∀ (n : ℕ), M →ₗ[R] N ⧸ (I ^ n • ⊤ : Submodule R N))
```

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +203,259 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+  rfl
+
+@[simp]
+lemma lift_eval_apply (f : ∀ (n : ℕ), M →ₗ[R] N ⧸ (I ^ n • ⊤ : Submodule R N))
```
```suggestion
lemma lift_apply (f : ∀ (n : ℕ), M →ₗ[R] N ⧸ (I ^ n • ⊤ : Submodule R N))
```
(or `lift_apply_apply` or `eval_lift_apply`)

erdOne reviewed (approved):
Thanks for the quick response!
Some more nitpicks:

chrisflav commented:
Besides applying your suggestions, I added an `ext` lemma for `AdicCauchySequence` and renamed the `val_add` (etc.) lemmas for `AdicCauchySequence` to `add_apply` (etc.) (in accordance with `Pi.add_apply`).

erdOne commented on Mathlib/RingTheory/AdicCompletion/Basic.lean:
```diff
@@ -201,67 +203,279 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
+
+/-- To show a statement about an element of `adicCompletion I M`, it suffices to check it
+on Cauchy sequences. -/
+theorem inductionOn {p : AdicCompletion I M → Prop} (x : AdicCompletion I M)
```
```suggestion
theorem induction_on {p : AdicCompletion I M → Prop} (x : AdicCompletion I M)
```

erdOne commented:
Thanks!
maintainer merge

riccardobrasca commented:
Thanks!

bors merge
